### ISSUE TITLE
Make `from_config.rs` modular

### ISSUE THREAD

**Issue by tusharmath** [tailcallhq/tailcall#384]
**Problem**
`from_config.rs` is a large file with a lot of convoluted logic inside of it. It makes it hard to refactor or reason about.

**Solution**
Split from_config into smaller module with separate files for easier maintenance.

**Technical Requirements**
- Split from_config as transformers of type `blueprint::Transform<&Config, Blueprint>`.
- Create transformers and files for each [operator].
- Combine all the transformers inside of `from_config.rs` using the `+` operator on `Transform`
- All existing tests should pass.

[operator]: https://tailcall.run/docs/intro/operators/

**Comment by akansh-bansal** [comment]
@Hitesh-sisara and @Nike682631 good to see you guys also attempting this, I've already created a few files related to transformer modules in my draft PR. If you're interested, let's collaborate on this.

**Comment by 0x61nas** [comment]
> @Hitesh-sisara and @Nike682631 good to see you guys also attempting this, I've already created a modules in my draft PR. If you're interested, let's collaborate on this.

I checked the PR, and it appears that there was no activity for about 5 days. I would love to know if you are still working on it.

**Comment by tusharmath** [comment]
@0x61nas Absolutely! The bounty is up for grabs, all the best :)

**Comment by mainrs** [comment]
Is this still needed? There has been a PR that did great work. What where the remaining issues that need to be solved? I'd like to work on this.

**Comment by tusharmath** [comment]
There was some work that we started on with, however it's not completed. We want it to be split into smaller modules (even if it's one file) and compose them together using `TryFold`.

I think the following structure could work — 

```
from_config/
├── mod.rs
├── server.rs
├── upstream.rs
└── graphql/
    ├── mod.rs
    ├── schema.rs
    ├── union.rs
    └── types/
        ├── mod.rs
        ├── const.rs
        ├── http.rs
        ├── inline.rs
        ├── javascript.rs
        └── modify.rs
```

Let me know if you'd be interested in pursuing this, I can assign it to you. Just a head's up, this isn't for the faint hearted, as you will need to deal with a ton of what we call occupational hazard - AKA Git Conflicts!

**Comment by mainrs** [comment]
Just a heads-up: I'm done with `upstream` and `server`, taking some inspiration from the previous PR for changing the `Valid` struct to a type alias. 

I can see how making changes results in a lot of merge conflicts. I do want to recommend splitting the code into different crates as a larger refactor approach. I think it would be a good idea. As of now, some parts are still coupled together that can definitely be independent from each other.

**Comment by mainrs** [comment]
@tusharmath Could you explain to me this part of the code? What is the idea behind this? What is the purpose of the `Type` struct and how does it relate to `@modify` for example? Why does the directive has to check the type of the specified field under the `name` attribute ?

https://github.com/tailcallhq/tailcall/blob/4bd426eb6ae79e6981e2dd3c3694bf792bef6c82/src/blueprint/from_config.rs#L501-L523

**Comment by tusharmath** [comment]
>What is the idea behind this? 

This is where we check if we can use modify operator and then apply it. Learn more about the modify operator here — https://tailcall.run/docs/guides/operators/#modify

>What is the purpose of the Type struct and how does it relate to @modify for example?
Type struct is a representation of any "type of data" that is exposed via an API. Type could be an interface also, meaning it's not real and other types implement it. In this case, if an interface has a particular field, which is being implemented in a particular type, the field can't be modified.

> Why does the directive has to check the type of the specified field under the name attribute ?
For modify to work correctly it needs to know if the field is only defined in the type or if it's coming from the implemented interface.

**Comment by mainrs** [comment]
After having worked a little bit more on it I have to admit that it takes quite some time to solve this issue. I have done some work on it, and I think the code I have is a good framework to build on top of.

It splits the code base into smaller chunks and makes it super easy to extend.

But sadly I have no time left to work on this. I pushed my code to https://github.com/mainrs/tailcall/tree/refactor/from_config.

The only thing left to do is to merge all the transformations together. Missing transformations are the ones for GraphQL objects. In this transformation one has to also apply all the transformations for the custom directives like `http` and such. 

I will stop working on this, since I reached the time limit I set myself. Hopefully the next person can use my code and finish this refactor!
